### PR TITLE
State Transfer bugs fixes cont’d:

### PR DESCRIPTION
**Pull request by toly-kournik**
1. Restarting replica during state transfer in process
   - Before first state transfer checkpoint reached - don't load reserved pages
   - loadReplicaData() - fix index in seqNumWinArray

2. ClientsManager::allocateNewReplyMsgAndWriteToStorage()
   - allow equality in assert in favour of crash recovery

3. ReplicaImpl
  -  not to use setFetchingState() as no one eventually reads this value as it’s set in state transfer functionality.

4. Re-enable skvbc_persistence_tests

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -1384,12 +1384,6 @@ void ReplicaImp::onMessage(CheckpointMsg *msg) {
   if (askForStateTransfer) {
     LOG_INFO_F(GL, "call to startCollectingState()");
 
-    if (ps_) {
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ReplicaImp.cpp]: If we are no longer using `setFetchingState` we should remove it and `getFetchingState` completely in a future commit.

```
andrewstone@ubuntu:~/concord-bft$ grep getFetchingState bftengine/src/bftengine/*
bftengine/src/bftengine/DebugPersistentStorage.cpp:bool DebugPersistentStorage::getFetchingState() {
bftengine/src/bftengine/DebugPersistentStorage.hpp:  bool getFetchingState() override;
bftengine/src/bftengine/PersistentStorage.hpp:  virtual bool getFetchingState() = 0;
bftengine/src/bftengine/PersistentStorageImp.cpp:bool PersistentStorageImp::getFetchingState() {
bftengine/src/bftengine/PersistentStorageImp.hpp:  bool getFetchingState() override;
andrewstone@ubuntu:~/concord-bft$ grep setFetchingState bftengine/src/bftengine/*
bftengine/src/bftengine/DebugPersistentStorage.cpp:void DebugPersistentStorage::setFetchingState(bool state) {
bftengine/src/bftengine/DebugPersistentStorage.hpp:  void setFetchingState(bool state) override;
bftengine/src/bftengine/PersistentStorage.hpp:  virtual void setFetchingState(bool state) = 0;
bftengine/src/bftengine/PersistentStorageImp.cpp:  setFetchingStateInternal(fetchingState_);
bftengine/src/bftengine/PersistentStorageImp.cpp:void PersistentStorageImp::setFetchingStateInternal(uint8_t state) {
bftengine/src/bftengine/PersistentStorageImp.cpp:void PersistentStorageImp::setFetchingState(bool state) {
bftengine/src/bftengine/PersistentStorageImp.cpp:  setFetchingStateInternal(state);
bftengine/src/bftengine/PersistentStorageImp.hpp:  void setFetchingState(bool state) override;
bftengine/src/bftengine/PersistentStorageImp.hpp:  void setFetchingStateInternal(uint8_t state);
bftengine/src/bftengine/ReplicaImp.cpp:      ps_->setFetchingState(true);
bftengine/src/bftengine/ReplicaImp.cpp:    ps_->setFetchingState(false);
bftengine/src/bftengine/ReplicaImp.cpp:      ps_->setFetchingState(true);
```

File: bftengine/src/bftengine/ReplicaLoader.cpp [hunk 2]
```diff
@@ -337,7 +337,7 @@ ReplicaLoader::ErrorCode loadReplicaData(shared_ptr<PersistentStorage> p, Loaded
       Verify(d.executedSeqNum > ld.lastStableSeqNum, InconsistentErr);
       Verify(d.executedSeqNum <= ld.lastStableSeqNum + kWorkWindowSize, InconsistentErr);
 
-      uint64_t idx = d.executedSeqNum - ld.lastStableSeqNum;
+      uint64_t idx = d.executedSeqNum - ld.lastStableSeqNum - 1;
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ReplicaLoader.cpp]: Why was this change made?

if `d.executedSeqNum == 0 && ld.lastStableSeqNum == 0` then `idx` will underflow become UINT64_MAX and the Assert below will fire.

Comment (andrewjstone) [reply bftengine/src/bftengine/ReplicaLoader.cpp]: Nevermind the last part of this. I see from the surrounding code (not shown in the PR) that this is safe as `d.executedSeqNum` must be greater than `ld.lastStableSeqNum.`

I'm still curious as to why this change is needed though. I just want to understand. I don't doubt it is correct.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 1 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaImp.cpp]: I agree

File: bftengine/src/bftengine/ReplicaLoader.cpp (hunk 2 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaLoader.cpp]: It just what we saw during the debugging. @guyg8  confirmed.

**Pull request merged**